smt/pkg/smt/witness_test.go

```diff
@@ -101,6 +103,90 @@ func findNode(t *testing.T, w *trie.Witness, addr libcommon.Address, storageKey
 	return nil
 }
 
+// TestWitnessToSMT tests that the SMT built from a witness matches the original SMT
+func TestWitnessToSMT(t *testing.T) {
+	smtTrie, rl := prepareSMT(t)
+
+	witness, err := smt.BuildWitness(smtTrie, rl, context.Background())
+	if err != nil {
+		t.Errorf("error building witness: %v", err)
+	}
+
+	newSMT, err := smt.BuildSMTfromWitness(witness)
+	if err != nil {
+		t.Errorf("error building SMT from witness: %v", err)
+	}
+
+	root, err := newSMT.Db.GetLastRoot()
+	if err != nil {
+		t.Errorf("error getting last root: %v", err)
+	}
+
+	// newSMT.Traverse(context.Background(), root, func(prefix []byte, k utils.NodeKey, v utils.NodeValue12) (bool, error) {
+	// 	fmt.Printf("[After] path: %v, hash: %x\n", prefix, libcommon.BigToHash(k.ToBigInt()))
+	// 	return true, nil
+	// })
+
+	expectedRoot, err := smtTrie.Db.GetLastRoot()
+	if err != nil {
+		t.Errorf("error getting last root: %v", err)
+	}
+
+	// assert that the roots are the same
+	if expectedRoot.Cmp(root) != 0 {
+		t.Errorf(fmt.Sprintf("SMT root mismatch, expected %x, got %x", expectedRoot.Bytes(), root.Bytes()))
+	}
+}
+
+// TestWitnessToSMTStateReader tests that the SMT built from a witness matches the state
+func TestWitnessToSMTStateReader(t *testing.T) {
+	smtTrie, rl := prepareSMT(t)
+
+	expectedRoot, err := smtTrie.Db.GetLastRoot()
+	if err != nil {
+		t.Errorf("error getting last root: %v", err)
+	}
+
+	witness, err := smt.BuildWitness(smtTrie, rl, context.Background())
+	if err != nil {
+		t.Errorf("error building witness: %v", err)
+	}
+
+	newSMT, err := smt.BuildSMTfromWitness(witness)
+	if err != nil {
+		t.Errorf("error building SMT from witness: %v", err)
+	}
+	root, err := newSMT.Db.GetLastRoot()
+	if err != nil {
+		t.Errorf("error building SMT from witness: %v", err)
+	}
+
+	if expectedRoot.Cmp(root) != 0 {
+		t.Errorf(fmt.Sprintf("SMT root mismatch, expected %x, got %x", expectedRoot.Bytes(), root.Bytes()))
+	}
+
+	contract := libcommon.HexToAddress("0x71dd1027069078091B3ca48093B00E4735B20624")
+
+	expectedAcc, _ := smtTrie.ReadAccountData(contract)
+	newAcc, _ := newSMT.ReadAccountData(contract)
+
+	expectedAccCode, _ := smtTrie.ReadAccountCode(contract, 0, expectedAcc.CodeHash)
+	newAccCode, _ := newSMT.ReadAccountCode(contract, 0, newAcc.CodeHash)
+	expectedAccCodeSize, _ := smtTrie.ReadAccountCodeSize(contract, 0, expectedAcc.CodeHash)
+	newAccCodeSize, _ := newSMT.ReadAccountCodeSize(contract, 0, newAcc.CodeHash)
+
```

Comment:
It will be great if we can test retrieving the value from a storage slot as well.

Reply:
Done ✅